skrub/_datetime_encoder.py

```diff
@@ -18,62 +21,494 @@
+        - If a float or integer, origin is the millisecond difference
+          relative to 1970-01-01.
+    cache : bool, default True
+        If :const:`True`, use a cache of unique, converted dates to apply the
+        datetime conversion. May produce significant speed-up when parsing
+        duplicate date strings, especially ones with timezone offsets. The cache
+        is only used when there are at least 50 values. The presence of
+        out-of-bounds values will render the cache unusable and may slow down
+        parsing.
+
+    Returns
+    -------
+    datetime
+        If parsing succeeded.
+        Return type depends on input (types in parenthesis correspond to
+        fallback in case of unsuccessful timezone or out-of-range timestamp
+        parsing):
+
+        - scalar: :class:`Timestamp` (or :class:`datetime.datetime`)
+        - array-like: :class:`DatetimeIndex` (or :class:`Series` with
+          :class:`object` dtype containing :class:`datetime.datetime`)
+        - Series: :class:`Series` of :class:`datetime64` dtype (or
+          :class:`Series` of :class:`object` dtype containing
+          :class:`datetime.datetime`)
+        - DataFrame: :class:`Series` of :class:`datetime64` dtype (or
+          :class:`Series` of :class:`object` dtype containing
+          :class:`datetime.datetime`)
+
+    Raises
+    ------
+    ParserError
+        When parsing a date from string fails.
+    ValueError
+        When another datetime conversion error happens. For example when one
+        of 'year', 'month', day' columns is missing in a :class:`DataFrame`, or
+        when a Timezone-aware :class:`datetime.datetime` is found in an array-like
+        of mixed time offsets, and ``utc=False``.
+
+    See Also
+    --------
+    :func:`pandas.to_datetime`
+    """
+    kwargs["errors"] = errors
+
+    # dataframe
+    if hasattr(X, "__dataframe__"):
+        return _to_datetime_dataframe(X, **kwargs)
+
+    # series, this attribute is available since Pandas 2.1.0
+    elif hasattr(X, "__column_consortium_standard__"):
+        return _to_datetime_series(X, **kwargs)
+
+    # 2d array
+    elif isinstance(X, Iterable) and np.asarray(X).ndim == 2:
+        X = _to_datetime_2d_array(np.asarray(X), **kwargs)
+        return np.vstack(X).T
+
+    # scalar or unknown type
+    return pd.to_datetime(X, **kwargs)
+
+
+def _to_datetime_dataframe(X, **kwargs):
+    """Dataframe specialization of ``_to_datetime_2d``.
+
+    Parameters
+    ----------
+    X : Pandas or Polars dataframe
+
+    Returns
+    -------
+    X : Pandas or Polars dataframe
+    """
+    _, px = get_df_namespace(X)
+    index = getattr(X, "index", None)
+    X_split = [X[col].to_numpy() for col in X.columns]
+    X_split = _to_datetime_2d(X_split, **kwargs)
+    X_split = {col: X_split[col_idx] for col_idx, col in enumerate(X.columns)}
+    X = pd.DataFrame(X_split, index=index)
+    # conversion is px is Polars, no-op if Pandas
+    return px.DataFrame(X)
```

Comment:
also it is not due to the skrub code but the conversion fails when timezones are involved:

```python
import polars as pl
from skrub import to_datetime

df = pl.DataFrame({"date": ["2023-10-12T17:36:50+01:00"]})
to_datetime(df) # pl.exceptions.ComputeError
```

but this seems to be more of a polars & pandas compatibility issue:

```python
import pandas as pd
import polars as pl

df = pd.DataFrame({"date": ["2023-10-12T17:36:50+02:00"]})
df["date"] = pd.to_datetime(df["date"])
pl.DataFrame(df)
```

I'll open an issue on the polars repo

Reply:
> could we build a polars dataframe directly instead of pandas first and then convert? AFAIK at the moment initializing a polars dataframe with a pandas one will make a copy

Polars is still an optional dependency, so I guess we can't

Reply:
> I'll open an issue on the polars repo

Thank you for this! Note that I haven't written tests for Polars yet, maybe in a subsequent PR to avoid obfuscating this one.

Reply:
Yes indeed!

Reply:
> I'll open an issue on the polars repo

I opened https://github.com/pola-rs/polars/issues/11774